File: javascript/browser/security/eval-detected.js
/**
 * Only report `eval` when we provide it with non-constant parameters.
 */

/**
 * Negative matches
 */

// ok:eval-detected
eval('var x = "static strings are okay";');

// ok:eval-detected
const constVar = "function staticStrings() { return 'static strings are okay';}";
eval(constVar);

// ok - const within another const
eval(`${constVar}`);

// ok - concatenating with another const okay
const secondConstVar = 'this is a const variable';
eval(constVar + secondConstVar);

/**
 * Positive Matches
 */

let dynamic = window.prompt() // arbitrary user input

// ruleid:eval-detected
eval(dynamic + 'possibly malicious code');

// ruleid:eval-detected
eval(`${dynamic} possibly malicious code`);

// ruleid:eval-detected
eval(dynamic.concat(''));

function evalSomething(something) {
    // ruleid:eval-detected
    eval(something);
}

File: python/lang/security/audit/eval-detected.py
# ok:eval-detected
eval("x = 1; x = x + 2")

# ok:eval-detected
eval(f"x = 1; x = x + 2")

blah = "import requests; r = requests.get('https://example.com')"
# ok:eval-detected
eval(blah)

dynamic = "import requests; r = requests.get('{}')"
# ruleid:eval-detected
eval(dynamic.format("https://example.com"))


def eval_something(something):
    # ruleid:eval-detected
    eval(something)


from something import eval

# ok:eval-detected
eval("something")

# ok:eval-detected
eval("somethin(){}")

# ok:eval-detected
eval(f"something()")

# ok:eval-detected
eval("")

# ok:eval-detected
eval(f"")

user_input = get_userinput()
# ruleid:eval-detected
eval(f"some_func({user_input})")

def eval_something(something):
    # ruleid:eval-detected
    eval(f"some_func({{{something}}})")
